snakemake/common/plugin.py

```diff
@@ -0,0 +1,132 @@
+
+    Raises a PluginException if this is not true.
+
+    Also takes a checks dictionary with a function and an optional exception.
+    The check should return None or False to pass.
+    """
+    package_name = cached_packages_distributions[mod.__name__][0]
+    if prefix:
+        if not package_name.startswith(prefix):
+            raise PluginException(
+                f"Plugin {package_name} is not distributed "
+                f"with a package starting with {prefix}."
+            )
+    for a in mod_attrs + extra_attrs:
+        if not hasattr(mod, a):
+            raise PluginException(
+                f"Plugin {package_name} does not provide attribute {a}."
+            )
+    try:
+        min_version(mod.__min_snakemake_version__)
+    except Exception:
+        raise PluginException(
+            "Plugin {mod.__name__} needs minimum Snakemake version {mod.__min_snakemake_version__}."
+        )
+
+
+def load_plugin(plugin_module, globals_dict, verify_func, alternative_mod_name=None):
+    """Load plugins into the globals dict.
+    Additionally some checks are performed on the plugins.
+    The module is bround with the name and the lowercase of the name.
+
+    plugin_module: module of the plugin
+    globals_dict: result of globals to the set module on
+    verify_func: function that verifies plugin or throw PluginException if wrong
+    alternative_mod_name: source name on the plugin module that provides the name
+        to assign the module on the globals_dict.
+
+    returns True if plugin is loaded.
+    """
+    plugin_name = plugin_module.__name__
+    package_name = cached_packages_distributions[plugin_name][0]
+    if plugin_module.__name__ in globals_dict:
+        logger.debug("Plugin {package_name} already loaded.")
+        return True
+
+    try:
+        verify_func(plugin_module)
+    except PluginException as e:
+        logger.warning(f"Plugin {package_name} incorrect: {e}")
+        return False
+
+    snakemake_submodule_name = getattr(
+        plugin_module,
+        alternative_mod_name or "snakemake_submodule_name",
+        plugin_name,
+    )
+    globals_dict[snakemake_submodule_name] = plugin_module
+    globals_dict[snakemake_submodule_name.lower()] = plugin_module
+    return True
+
+
+def internal_submodules(paths):
+    submodules = []
+    for remote_submodule in pkgutil.iter_modules(paths):
+        module_name = remote_submodule.name
+
+        module_path = (
+            path.join(remote_submodule.module_finder.path, module_name) + ".py"
+        )
+
+        spec = importlib.util.spec_from_file_location(module_name, module_path)
+        module = importlib.util.module_from_spec(spec)
+
+        try:
+            sys.modules[module_name] = module
+            spec.loader.exec_module(module)
+        except Exception as e:
+            logger.debug(f"Autoloading {module_name} failed: {e}")
+            continue
+
+        submodules(module)
```

Comment:
isn't submodules defined as a list above?

Reply:
yes, it should be submodules.append

that's a bug. fixed no.